libs/gltfio/src/StbProvider.cpp

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright (C) 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gltfio/TextureProvider.h>
+
+#include <string>
+#include <vector>
+
+#include <utils/JobSystem.h>
+#include <utils/Log.h>
+
+#include <filament/Engine.h>
+#include <filament/Texture.h>
+
+#include <stb_image.h>
+
+using namespace filament;
+using namespace utils;
+
+using std::string;
+
+namespace gltfio {
+
+class StbProvider : public TextureProvider {
+public:
+    StbProvider(Engine* engine);
+    ~StbProvider();
+
+    Texture* pushTexture(const uint8_t* data, size_t byteCount,
+            const char* mimeType, uint64_t flags) final;
+
+    Texture* popTexture() final;
+
+    void updateQueue() final;
+
+    const char* getFailureMessage() const final { return mErrorMessage.c_str(); }
+
+    void waitForCompletion() final;
+
+    size_t getPushedCount() const final { return mPushedCount; }
+
+    size_t getPoppedCount() const final { return mPoppedCount; }
+
+    size_t getDecodedCount() const final { return mDecodedCount; }
+
+private:
+    enum class TextureState {
+        DECODING, // Texture has been pushed, mipmap levels are not yet complete.
+        READY,    // Mipmap levels are available but texture has not been popped yet.
+        POPPED,   // Client has popped the texture from the queue.
+    };
+
+    struct TextureInfo {
+        Texture* texture;
+        TextureState state;
+        std::atomic<intptr_t> decodedTexelsBaseMipmap;
+        std::vector<uint8_t> sourceBuffer;
+        JobSystem::Job*  decoderJob;
+    };
+
+    static const intptr_t DECODING_ERROR = 0x1;
```

Comment:
Maybe a constant for `DECODING_NOTSTARTED = 0x0;` or something. I think that'd help readability.

Reply:
Good idea, I'll do this.